c10/core/TensorImpl.cpp

```diff
@@ -227,119 +227,171 @@ void TensorImpl::HandleResize() {
   }
 }
 
-bool TensorImpl::compute_contiguous() const {
+template <typename T>
+bool_is_contiguous _compute_contiguous(
+    ArrayRef<T> sizes,
+    ArrayRef<T> strides,
+    T numel) {
   bool is_contiguous = true;
-  if (is_empty())
-    return is_contiguous;
-  int64_t z = 1;
-  for (int64_t d = dim() - 1; d >= 0; d--) {
-    const auto size_d = sizes_and_strides_.size_at_unchecked(d);
+  if (numel == 0)
+    return bool_is_contiguous(is_contiguous);
+  T z = 1;
+  // NB: make sure we do signed arithmetic
+  for (int64_t d = int64_t(sizes.size()) - 1; d >= 0; d--) {
+    const auto size_d = sizes[d];
     if (size_d != 1) {
-      if (sizes_and_strides_.stride_at_unchecked(d) == z) {
+      if (strides[d] == z) {
         z *= size_d;
       } else {
         is_contiguous = false;
         break;
       }
     }
   }
-  return is_contiguous;
+  return bool_is_contiguous(is_contiguous);
 }
 
-bool TensorImpl::compute_channels_last_contiguous_2d() const {
+// NB: intentionally bypass the normal accessors; we always want to be
+// consistent with what is actually stored on the struct
+#define COMPUTE_WITH_SIZES_STRIDES_NUMEL(TEMPLATE)                            \
+  (has_symbolic_sizes_strides_                                                \
+       ? TEMPLATE<c10::SymInt>(                                               \
+             extra_meta_->sizes_, extra_meta_->strides_, extra_meta_->numel_) \
+       : TEMPLATE<int64_t>(                                                   \
+             sizes_and_strides_.sizes_arrayref(),                             \
+             sizes_and_strides_.strides_arrayref(),                           \
+             numel_))
+
+#define COMPUTE_WITH_SIZES_STRIDES(TEMPLATE)                               \
+  (has_symbolic_sizes_strides_                                             \
+       ? TEMPLATE<c10::SymInt>(extra_meta_->sizes_, extra_meta_->strides_) \
+       : TEMPLATE<int64_t>(                                                \
+             sizes_and_strides_.sizes_arrayref(),                          \
+             sizes_and_strides_.strides_arrayref()))
+
+bool_is_contiguous TensorImpl::compute_contiguous() const {
+  return COMPUTE_WITH_SIZES_STRIDES_NUMEL(_compute_contiguous);
+}
+
+template <typename T>
+bool_is_channels_last_contiguous _compute_channels_last_contiguous_2d(
```

Comment:
just out of curiosity, what problem is the use of strong types solving here?  accidentally putting the wrong bools in the wrong slots? or implicit conversions?

Reply:
Putting bools in wrong slots.